schema/src/Mu/Schema/Quasi.hs

```diff
@@ -1,102 +1,171 @@
+avro :: QuasiQuoter
+avro =
+  QuasiQuoter
+    (const $ fail "cannot use as expression")
+    (const $ fail "cannot use as pattern")
+    schemaFromAvroString
+    (const $ fail "cannot use as declaration")
 
 -- | Imports a protocol buffer definition written
 --   in-line as a 'Schema'.
 protobuf :: QuasiQuoter
-protobuf = QuasiQuoter (const $ fail "cannot use as expression")
-                       (const $ fail "cannot use as pattern")
-                       schemaFromProtoBufString
-                       (const $ fail "cannot use as declaration")
+protobuf =
+  QuasiQuoter
+    (const $ fail "cannot use as expression")
+    (const $ fail "cannot use as pattern")
+    schemaFromProtoBufString
+    (const $ fail "cannot use as declaration")
+
+-- | Imports an avro definition from a file as a 'Schema'.
+avroFile :: QuasiQuoter
+avroFile = quoteFile avro
 
 -- | Imports a protocol buffer definition from a file
 --   as a 'Schema'.
 protobufFile :: QuasiQuoter
 protobufFile = quoteFile protobuf
 
+schemaFromAvroString :: String -> Q Type
+schemaFromAvroString s =
+  case decode (encode s) of
+    Nothing -> fail "could not parse avro spec!"
+    Just p  -> schemaFromAvroType p
+
+schemaFromAvroType :: A.Type -> Q Type
+schemaFromAvroType sch =
+  case sch of
+    A.Null -> [t|'TPrimitive 'TNull|]
+    A.Boolean -> [t|'TPrimitive Bool|]
+    A.Int -> [t|'TPrimitive Int32|]
+    A.Long -> [t|'TPrimitive Int64|]
+    A.Float -> [t|'TPrimitive Float|]
+    A.Double -> [t|'TPrimitive Double|]
+    A.Bytes -> [t|'TPrimitive B.ByteString|]
+    A.String -> [t|'TPrimitive T.Text|]
+    A.Array item -> [t|'TList $(schemaFromAvroType item)|]
+    A.Map values -> [t|'TMap $(schemaFromAvroType values)|]
+    A.NamedType _ -> fail "named types are not currently supported"
+    A.Record name _ _ _ fields ->
+      [t|'DRecord $(textToStrLit $ A.baseName name) '[] $(typesToList <$>
+                                                          mapM
+                                                            avroFieldToType
+                                                            fields)|]
+      where avroFieldToType :: A.Field -> Q Type
+            avroFieldToType field =
+              [t|'FieldDef $(textToStrLit $ A.fldName field) '[] $(schemaFromAvroType $
+                                                                   A.fldType
+                                                                     field)|]
+    A.Enum name _ _ symbols ->
+      [t|'DEnum $(textToStrLit $ A.baseName name) '[] $(typesToList <$>
+                                                        mapM
+                                                          avChoiceToType
+                                                          (toList symbols))|]
+      where avChoiceToType :: T.Text -> Q Type
+            avChoiceToType c = [t|'ChoiceDef $(textToStrLit c) '[]|]
+    A.Union _ -> fail "unions are not currently supported"
```

Comment:
Check the special case of union of `Null` and another type, to be map to an optional.

Reply:
Is this implementation correct? 🤔 https://github.com/higherkindness/mu-haskell/pull/18/commits/0a229fc75d98fc29367a1564a9d239fe5f40ae2f#diff-df426cfe6cf9b36c0492ab73d0d05c05R88